File: TimeCardClientAndroid/app/src/main/java/jp/uhimania/timecardclientandroid/data/DateExtension.kt
package jp.uhimania.timecardclientandroid.data

import jp.uhimania.japanholidays.Holidays
import java.util.Calendar
import java.util.Date

fun Date.year(): Int {
    val calendar = Calendar.getInstance()
    calendar.time = this
    return calendar.get(Calendar.YEAR)
}

fun Date.month(): Int {
    val calendar = Calendar.getInstance()
    calendar.time = this
    return calendar.get(Calendar.MONTH) + 1
}

fun Date.day(): Int {
    val calendar = Calendar.getInstance()
    calendar.time = this
    return calendar.get(Calendar.DAY_OF_MONTH)
}

fun Date.hour(): Int {
    val calendar = Calendar.getInstance()
    calendar.time = this
    return calendar.get(Calendar.HOUR_OF_DAY)
}

fun Date.minute(): Int {
    val calendar = Calendar.getInstance()
    calendar.time = this
    return calendar.get(Calendar.MINUTE)
}

fun Date.second(): Int {
    val calendar = Calendar.getInstance()
    calendar.time = this
    return calendar.get(Calendar.SECOND)
}

fun Date.isHoliday(): Boolean {
    val calendar = Calendar.getInstance()
    calendar.time = this
    val dayOfWeek = calendar.get(Calendar.DAY_OF_WEEK)
    return dayOfWeek == Calendar.SUNDAY || dayOfWeek == Calendar.SATURDAY || Holidays.isHoliday(this)
}

File: TimeCard/Extensions/DateExtension.swift
//
//  DateExtension.swift
//  TimeCard
//
//  Created by uhimania on 2025/10/02.
//

import Foundation
import JapanHolidays

extension Date {
    var year: Int {
        Calendar.current.component(.year, from: self)
    }
    
    var month: Int {
        Calendar.current.component(.month, from: self)
    }
    
    var day: Int {
        Calendar.current.component(.day, from: self)
    }
    
    var hour: Int {
        Calendar.current.component(.hour, from: self)
    }
    
    var minute: Int {
        Calendar.current.component(.minute, from: self)
    }
    
    var second: Int {
        Calendar.current.component(.second, from: self)
    }
    
    var weekDay: String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "ja_JP")
        formatter.dateFormat = "E"
        return formatter.string(from: self)
    }
    
    func isHoliday() -> Bool {
        let weekday = Calendar.current.component(.weekday, from: self)
        return weekday == 1 || weekday == 7 || Holidays.isHoliday(self)
    }
    
    struct DayWithWeekdayFormatStyle: Foundation.FormatStyle {
        func format(_ value: Date) -> String {
            let formatter = DateFormatter()
            formatter.dateFormat = "dd(E)"
            formatter.locale = Locale(identifier: "ja_JP")
            return formatter.string(from: value)
        }
    }
}

extension FormatStyle where Self == Date.DayWithWeekdayFormatStyle {
    static var dayWithWeekday: Self { .init() }
}
